### PR TITLE
Hooks refactor

### PR DESCRIPTION
**Pull request by tbrannam**
<h1>Table of Contents</h1>

- [Description](#description)
- [Motivation and Context](#motivation-and-context)
- [How Has This Been Tested?](#how-has-this-been-tested)
- [Screenshots (if appropriate):](#screenshots-if-appropriate)
- [Types of changes](#types-of-changes)
- [Checklist:](#checklist)

## Description

Introduce hooks support
Refactor CoreExperiment to use hooks
Update tests.
Fix slow tests

## Motivation and Context

Ready support for react 17.x 
Add hooks support for React Functional Components

## How Has This Been Tested?

Updated tests related to params changes for core experiment.
Added specific tests for hook, useExperiment

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: test/browser/weighted.test.jsx [hunk 1]
```diff
@@ -55,7 +54,8 @@ describe('Weighted Experiment', () => {
     });
 
     for (let i = 0; i < 1000; i++) {
-      mount(<App />);
+      const container = mount(<App />);
+      container.unmount();
```
Comment (tbrannam) [inline test/browser/weighted.test.jsx:58]: looks like memory pressure was what was making this test take a long time to complete

**Comment by moretti** [comment]
Hey @tbrannam, looks like an amazing work, I'll have deep look tomorrow.
Having support for hooks is great, I think we'll have to create a major release, since we're no longer supporting `react@0.14` and `react@15`, but it should be fine, we can potentially add it to README.

I'm sorry about the late reply (currently drowning in reviews 😅), thank you for submitting this pull request! 🎉

**Comment by tbrannam** [comment]
It's all good - I think I did everything I wanted to do - just added tests for hooks.

File: README.md [hunk 2]
```diff
@@ -94,18 +94,52 @@ yarn add @marvelapp/react-ab-test
...
 import { Experiment, Variant, emitter } from '@marvelapp/react-ab-test';
 
 class App extends Component {
+  experimentRef = React.createRef();
+
+  onButtonClick(e) {
+    this.experimentRef.current.win();
+
   onButtonClick(e) {
     this.refs.experiment.win();
```
Comment (moretti) [inline README.md]: Thank you for removing string refs!

The second `onButtonClick` is duplicated and should be removed.

---

There two comments are still referencing `this.refs.experiment`:

https://github.com/marvelapp/react-ab-test/blob/756a87ee2e187fd0016bb2299cdb530d8abc50c3/README.md#L161-L165

https://github.com/marvelapp/react-ab-test/blob/756a87ee2e187fd0016bb2299cdb530d8abc50c3/README.md#L233-L236

File: src/CoreExperiment.jsx [hunk 3]
```diff
@@ -1,83 +1,46 @@
...
-  render() {
-    return this.state.variants[this.state.value] || null;
-  }
+  return selectVariant(variants, []);
 };
+
+CoreExperiment.propTypes = {
+  name: PropTypes.string.isRequired,
+  userIdentifier: PropTypes.string,
+  defaultVariantName: PropTypes.string,
```
Comment (moretti) [inline src/CoreExperiment.jsx:43]: The original component did not include `children`:
```.js
children: PropTypes.node,
```
can you add it to `propTypes`?

File: src/hooks.jsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import { useEffect, useState } from "react";
+import emitter from "./emitter";
+
+const selectVariant = (currentVariant) => (variants, fallback) => {
+  if (Object.keys(variants).indexOf(currentVariant) !== -1) {
```
Comment (moretti) [inline src/hooks.jsx]: I think we can use `variants.hasOwnProperty(currentVariant)` or `currentVariant in variants` here? 🤔

File: test/browser/emitter.test.jsx [hunk 5]
```diff
@@ -1,9 +1,10 @@
 import React, { Component } from 'react';
-import ReactDOM from 'react-dom';
 import UUID from 'uuid/v4';
 import { mount } from 'enzyme';
+import { act } from "@testing-library/react";
```
Comment (moretti) [inline test/browser/emitter.test.jsx]: I think we can use the built-in function from `ReactTestUtils` here?
```.js
import { act } from from 'react-dom/test-utils';
````
https://reactjs.org/docs/test-utils.html#act

I had a quick look at https://github.com/testing-library/react-testing-library/blob/8feb318b68d778c7fdd07337b66bf34e8e201f2b/src/act-compat.js
and I believe it's just a polyfill for react < 16.8.0
🤔

File: test/browser/weighted.test.jsx (hunk 1 above)
Comment (moretti) [reply test/browser/weighted.test.jsx:58]: Nice catch! 🎉

File: package.json [hunk 6]
```diff
@@ -11,20 +11,21 @@
...
   "main": "index.js",
-  "version": "2.3.0",
+  "version": "3.0.0",
   "description": "A/B testing React components and debug tools. Isomorphic with a simple, universal interface. Well documented and lightweight. Tested in popular browsers and Node.js. Includes helpers for Mixpanel and Segment.com.",
   "directories": {
     "test": "test"
   },
   "peerDependencies": {
-    "react": ">=0.14.0 <17.0.0"
+    "react": ">=16.3.0 <17.0.0"
```
Comment (moretti) [inline package.json]: Hooks were released with `react@^16.8.0`, so I think we need:
```
"react": ">=16.8.0 <17.0.0"
```

**Review by moretti — COMMENTED** [review]
This is really great work! 🎉
I left a couple of suggestions

**Comment by tbrannam** [comment]
Think I've addressed all the outstanding issues.    

One thing that I wonder about revisiting is the need to preregister variants.   This is due to the Hook not having any reference to the possible list of variants like the Components would have.    I like the hook interface now, but it might work if somehow the variants were passed to the hook as well as an alternative calling pattern.

Or maybe it isn't too bad the way it is?

**Comment by tbrannam** [comment]
friendly bump

**Comment by moretti** [comment]
I had another deep look at source code and I think it looks really good.

One way to solve the last problem that you mentioned would be to extend the `useExperiment` hook, eg. use an object for advance configuration:

```.js
useExperiment('experiment_name', { 
  variants: ['A', 'B'],
  defaultVariant: 'A',
  userIdentifier: 'pk_123', 
});
```

I think we could potentially merge this PR as it is, release a beta version of v3, test it for a few days and see how it goes, what do you think?

**Comment by tbrannam** [comment]
Sounds good
________________________________

**Comment by moretti** [comment]
Released in `3.0.0-beta.0` 🎉

**Pull request merged**